File: app/src/test/java/com/bas080/autosleepdroid/EventLoggerTest.java
package com.bas080.autosleepdroid;

import android.content.Context;
import android.text.Spanned;
import android.text.style.ForegroundColorSpan;
import androidx.test.core.app.ApplicationProvider;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.annotation.Config;

import java.util.List;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;

@RunWith(RobolectricTestRunner.class)
@Config(sdk = {34})
public class EventLoggerTest {

    private Context context;

    @Before
    public void setUp() {
        context = ApplicationProvider.getApplicationContext();
        EventLogger.clear(context);
    }

    @Test
    public void testLogLevelFormatting() {
        EventLogger.log(context, EventLogger.LEVEL_LOW, "Low importance message");
        EventLogger.log(context, EventLogger.LEVEL_NORMAL, "Normal importance message");
        EventLogger.log(context, EventLogger.LEVEL_HIGH, "High importance message");

        List<String> events = EventLogger.getEvents(context);
        assertEquals(3, events.size());

        // Check LEVEL_LOW line format and span color
        CharSequence formattedLow = EventLogger.formatColoredEvent(context, events.get(0));
        assertTrue(formattedLow instanceof Spanned);
        assertFalse(formattedLow.toString().contains("[L]"));
        assertTrue(formattedLow.toString().contains("Low importance message"));
        ForegroundColorSpan[] lowSpans = ((Spanned) formattedLow).getSpans(0, formattedLow.length(), ForegroundColorSpan.class);
        assertTrue(lowSpans.length >= 2);
        assertEquals(0xFF999999, lowSpans[0].getForegroundColor());
        assertEquals(0xFF888888, lowSpans[1].getForegroundColor());

        // Check LEVEL_NORMAL line format and span color
        CharSequence formattedNormal = EventLogger.formatColoredEvent(context, events.get(1));
        assertTrue(formattedNormal instanceof Spanned);
        assertFalse(formattedNormal.toString().contains("[N]"));
        assertTrue(formattedNormal.toString().contains("Normal importance message"));
        ForegroundColorSpan[] normalSpans = ((Spanned) formattedNormal).getSpans(0, formattedNormal.length(), ForegroundColorSpan.class);
        assertTrue(normalSpans.length >= 2);
        assertEquals(0xFF999999, normalSpans[0].getForegroundColor());
        assertEquals(0xFF444444, normalSpans[1].getForegroundColor());

        // Check LEVEL_HIGH line format and span color
        CharSequence formattedHigh = EventLogger.formatColoredEvent(context, events.get(2));
        assertTrue(formattedHigh instanceof Spanned);
        assertFalse(formattedHigh.toString().contains("[H]"));
        assertTrue(formattedHigh.toString().contains("High importance message"));
        ForegroundColorSpan[] highSpans = ((Spanned) formattedHigh).getSpans(0, formattedHigh.length(), ForegroundColorSpan.class);
        assertTrue(highSpans.length >= 2);
        assertEquals(0xFF999999, highSpans[0].getForegroundColor());
        assertEquals(0xFF000000, highSpans[1].getForegroundColor());
    }

    @Test
    @Config(qualifiers = "night")
    public void testLogLevelFormattingDarkMode() {
        EventLogger.log(context, EventLogger.LEVEL_LOW, "Low importance night message");
        EventLogger.log(context, EventLogger.LEVEL_NORMAL, "Normal importance night message");
        EventLogger.log(context, EventLogger.LEVEL_HIGH, "High importance night message");

        List<String> events = EventLogger.getEvents(context);
        assertEquals(3, events.size());

        CharSequence formattedLow = EventLogger.formatColoredEvent(context, events.get(0));
        ForegroundColorSpan[] lowSpans = ((Spanned) formattedLow).getSpans(0, formattedLow.length(), ForegroundColorSpan.class);
        assertEquals(0xFF888888, lowSpans[0].getForegroundColor());
        assertEquals(0xFFAAAAAA, lowSpans[1].getForegroundColor());

        CharSequence formattedNormal = EventLogger.formatColoredEvent(context, events.get(1));
        ForegroundColorSpan[] normalSpans = ((Spanned) formattedNormal).getSpans(0, formattedNormal.length(), ForegroundColorSpan.class);
        assertEquals(0xFF888888, normalSpans[0].getForegroundColor());
        assertEquals(0xFFDDDDDD, normalSpans[1].getForegroundColor());

        CharSequence formattedHigh = EventLogger.formatColoredEvent(context, events.get(2));
        ForegroundColorSpan[] highSpans = ((Spanned) formattedHigh).getSpans(0, formattedHigh.length(), ForegroundColorSpan.class);
        assertEquals(0xFF888888, highSpans[0].getForegroundColor());
        assertEquals(0xFFFFFFFF, highSpans[1].getForegroundColor());
    }
}

File: app/src/test/java/com/bas080/autosleepdroid/EventLoggerTest.kt
package com.bas080.autosleepdroid

import android.content.Context
import android.text.Spanned
import android.text.style.ForegroundColorSpan
import androidx.test.core.app.ApplicationProvider
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config

@RunWith(RobolectricTestRunner::class)
@Config(sdk = [34])
class EventLoggerTest {

    private lateinit var context: Context

    @Before
    fun setUp() {
        context = ApplicationProvider.getApplicationContext()
        EventLogger.clear(context)
    }

    @Test
    fun testLogLevelFormatting() {
        EventLogger.log(context, EventLogger.LEVEL_LOW, "Low importance message")
        EventLogger.log(context, EventLogger.LEVEL_NORMAL, "Normal importance message")
        EventLogger.log(context, EventLogger.LEVEL_HIGH, "High importance message")

        val events = EventLogger.getEvents(context)
        assertEquals(3, events.size)

        val formattedLow = EventLogger.formatColoredEvent(context, events[0])
        assertTrue(formattedLow is Spanned)
        assertFalse(formattedLow.toString().contains("[L]"))
        assertTrue(formattedLow.toString().contains("Low importance message"))
        val lowSpans = (formattedLow as Spanned).getSpans(0, formattedLow.length, ForegroundColorSpan::class.java)
        assertTrue(lowSpans.size >= 2)
        assertEquals(-0x666667, lowSpans[0].foregroundColor)
        assertEquals(-0x777778, lowSpans[1].foregroundColor)

        val formattedNormal = EventLogger.formatColoredEvent(context, events[1])
        assertTrue(formattedNormal is Spanned)
        assertFalse(formattedNormal.toString().contains("[N]"))
        assertTrue(formattedNormal.toString().contains("Normal importance message"))
        val normalSpans = (formattedNormal as Spanned).getSpans(0, formattedNormal.length, ForegroundColorSpan::class.java)
        assertTrue(normalSpans.size >= 2)
        assertEquals(-0x666667, normalSpans[0].foregroundColor)
        assertEquals(-0xbbbbbc, normalSpans[1].foregroundColor)

        val formattedHigh = EventLogger.formatColoredEvent(context, events[2])
        assertTrue(formattedHigh is Spanned)
        assertFalse(formattedHigh.toString().contains("[H]"))
        assertTrue(formattedHigh.toString().contains("High importance message"))
        val highSpans = (formattedHigh as Spanned).getSpans(0, formattedHigh.length, ForegroundColorSpan::class.java)
        assertTrue(highSpans.size >= 2)
        assertEquals(-0x666667, highSpans[0].foregroundColor)
        assertEquals(-0x1000000, highSpans[1].foregroundColor)
    }

    @Test
    @Config(qualifiers = "night")
    fun testLogLevelFormattingDarkMode() {
        EventLogger.log(context, EventLogger.LEVEL_LOW, "Low importance night message")
        EventLogger.log(context, EventLogger.LEVEL_NORMAL, "Normal importance night message")
        EventLogger.log(context, EventLogger.LEVEL_HIGH, "High importance night message")

        val events = EventLogger.getEvents(context)
        assertEquals(3, events.size)

        val formattedLow = EventLogger.formatColoredEvent(context, events[0])
        val lowSpans = (formattedLow as Spanned).getSpans(0, formattedLow.length, ForegroundColorSpan::class.java)
        assertEquals(-0x777778, lowSpans[0].foregroundColor)
        assertEquals(-0x555556, lowSpans[1].foregroundColor)

        val formattedNormal = EventLogger.formatColoredEvent(context, events[1])
        val normalSpans = (formattedNormal as Spanned).getSpans(0, formattedNormal.length, ForegroundColorSpan::class.java)
        assertEquals(-0x777778, normalSpans[0].foregroundColor)
        assertEquals(-0x222223, normalSpans[1].foregroundColor)

        val formattedHigh = EventLogger.formatColoredEvent(context, events[2])
        val highSpans = (formattedHigh as Spanned).getSpans(0, formattedHigh.length, ForegroundColorSpan::class.java)
        assertEquals(-0x777778, highSpans[0].foregroundColor)
        assertEquals(-0x1, highSpans[1].foregroundColor)
    }
}
